docs/architecture.md

```diff
@@ -4,213 +4,182 @@
+
+```typescript
+interface DealTracker {
+  // content of the offer
+  ferryCid: Link
+  // invocation and task CID to be able to go through receipts
+  invocation: Link
+  task: Link
+  // timestamp of invocation submitted
+  invoked: string
+  // timestamp of oldest piece
+  oldestPiece: string
+}
+```
+
+The `deal-queue` is the final stage of this multiple queue system. It tracks deal status until there is an update on the state of the deal to either `Approved` or `Rejected` by checking if a receipt exists. Once deal is settled, the ferry content is fetched and piece/aggregate mapping is written together with Deal status.
+
+![queues](./queue.svg)
+
+### Queues overview
+
+| name       | type     | batch | window | DLQ |
+|------------|----------|-------|--------|-----|
+| piece      | standard | 10000 | 300 s  | TBD |
+| batch      | FIFO     | 2     | 5 m    | TBD |
+| submission | FIFO     | 1     | 5 m    | TBD |
+| deal       | FIFO     | 10    | 5 m    | TBD |
+
+Other relevant notes:
+- with this approach we have an append only log where writes into the Database only happen when we have a deal in the very end. Moreover, it results in a quite small number of operations on the DB
+- while `w3up` CAR files can be limited to `4 GiB` to have a better utilization of Fil sector space, same does not currently happen with `pickup` (and perhaps other systems in the future). Designing assuming an upper limit is not a good way to go in this system.
+- current design enables us to quite easily support bigger deals in the future.
+- if an aggregate fails to land into a Storage Provider, the problematic piece(s) can be removed and a batch can be created created without that piece. This way, it can already be added to the `batch queue`
+
+Challenges/compromises:
+- where to hook alerts when things are getting delayed? we can hook alerts when requests to spade are failing, but will that be enough?
+  - perhaps we should handle failure tolerance (if needed) in `spade-proxy` where we keep track of items failing in a queue that we can hook up a Filecoin lite node
+
+## Data Stack
+
+The Data stack is responsible for the state of the w3filecoin.
+
+It keeps track of the received pieces, submitted aggregates, as well as in which aggregate a given piece is. In addition, it must keep the necessary state for the `buffer-queue` to operate, as well as the state of a given aggregate over time until a deal is fulfilled. It is worth mentioning, that keeping track of problematic pieces that could not be added to an aggregate should also be properly tracked.
+
+To achieve required state management, we will be relying on a S3 Bucket and a dynamoDB table.
+
+### Queries and Resources
+
+1. Check if piece is already in the pipeline (S3 Bucket)
```

Comment:
@Gozala looking into pricing for both models
- S3
  - $0.005 PUT, COPY, POST, LIST requests (per 1,000 requests)
  - $0.0004  GET, SELECT, and all other requests (per 1,000 requests)
- DynamoDB on demand capacity
  - $1.25 per million write request units
  - $0.25 per million read request units

Therefore, S3 has a cost for 1 M requests of $5 compared to $1.25

Swapping to dynamoDB